Integration ENS with IPFS

PhyrexTsai opened the pull request:
## Simple Summary
This implementation is to build a way to resolves ENS domains (.eth) to IPFS hash and redirects to url content.

## Motivation
MetaMask browser extension is great project helps users to interact with Ethereum, and furthermore can be more powerful by add content resolve by ENS.

## Specification
Resolve from ENS Public Resolver `getContent` hash to IPFS hash, and redirect url to `ipfs.infura.io`.

## Demo
![https://media.giphy.com/media/fs8Sh81IKp0GYFEMmF/giphy.gif](https://media.giphy.com/media/fs8Sh81IKp0GYFEMmF/giphy.gif)

## Video
https://www.dropbox.com/s/8g8y42z6w7e6bcj/metamask.mov?dl=0

MikeCheng1208 committed:
Intergrate ENS with IPFS

MikeCheng1208 committed:
Intergrate ENS with IPFS

MikeCheng1208 committed:
setTimeout bug fix

PhyrexTsai committed:
Update manifest.json

PhyrexTsai committed:
Rename files

PhyrexTsai committed:
Update lib

PhyrexTsai committed:
Update default provider

PhyrexTsai committed:
Update app

kumavis commented:
@PhyrexTsai this is awesome, please give us time to review it!

danfinlay commented on app/404.html:
```diff
@@ -0,0 +1,52 @@
+<body>
+    <div class="app">
+        <img src="./images/404.png" alt="">
+        <h2>Powered by <a href="https://www.portal.network/">Portal Network</a></h2>
```
I wasn't sure how I felt about this at first, but you know, getting credit for what you do seems pretty fair. I'm fine leaving it up, at least for a while. The same way we link to 409H for the phishing detector.

danfinlay commented on app/manifest.json:
```diff
@@ -59,14 +59,19 @@
-}
\ No newline at end of file
+  },
+  "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'"
```
Is there a way to not include the `unsafe-eval` permission?

danfinlay commented on app/manifest.json:
```diff
@@ -1,7 +1,7 @@
   "name": "__MSG_appName__",
   "short_name": "__MSG_appName__",
-  "version": "4.6.0",
+  "version": "4.6.2",
```
Please leave the version bumping to us.

danfinlay commented on app/manifest.json:
```diff
@@ -59,14 +59,19 @@
   "web_accessible_resources": [
-    "inpage.js"
+    "inpage.js",
+    "*"
```
I don't think we want to make *all* resources web-accessible. This seems like a potential privacy breech.

I think you should specifically list the files needed to display the page statuses (html, images).

danfinlay commented on app/scripts/lib/ipfsContent.js:
```diff
@@ -0,0 +1,36 @@
+
+            resolver.resolve(name, provider).then(ipfsHash => {
+                clearTimeout(clearTime)
+                let url = 'https://gateway.ipfs.io/ipfs/' + ipfsHash
```
Are we sure we want to hit IPFS gateway and not Infura?

Sending this comment to Infura for comment.

PhyrexTsai replied:
Sure, I have changed to the `ipfs.infura.io`

danfinlay commented on app/scripts/lib/resolver.js:
```diff
@@ -0,0 +1,71 @@
+function getProvider (type) {
+  switch (type) {
+    case 'mainnet':
+      return 'https://mainnet.infura.io/'
```
I would rather we use the user's provider, provided by MetaMask, to reduce our number of blockchain connections, but for this feature, I can see the benefit of this simplicity, there aren't other chains w/ the ENS resolver anyway, this _could_ be optimized later...

danfinlay reviewed (commented):
Just commenting for now, there are a few other opinions I'd like to get on parts of this.

PhyrexTsai committed:
Update manifest.json

PhyrexTsai committed:
Update ipfs to infura

egalano commented:
@danfinlay you can use https://ipfs.infura.io for this. We've put a lot of work into preparing for additional traffic for IPFS.

PhyrexTsai committed:
Update 404

PhyrexTsai committed:
Merge branch 'develop' into portal-metamask

PhyrexTsai committed:
Update 404.html

danfinlay commented:
@PhyrexTsai this would be very nicely accompanied by:

- A tutorial of how to post sites in this way.
- Some sample sites that work with this. The gif goes by quickly, we want to be able to show it off in full force when it comes out.
- A blog post for when we publish. I'll draft this one, but offer you a review before I post it.

danfinlay commented:
I've verified this works with `phyrextsai.eth`, it looks good, sorry for taking so long to try.

When I got it working, at first it usually didn't work, so I think it maybe only works with new tabs after installation, not sure why it was intermittently not working on first uses.

I mostly just want a guide on creating compatible sites, so others can build on this as soon as we release.

danfinlay reviewed (dismissed):
Works alright, has some issues with first time usage, and it doesn't save user's typed items, but does work with links like http://phyrextsai.eth, so that's cool, a blog post will really make this valuable to the community, but that doesn't need to block inclusion.

PhyrexTsai commented:
@danfinlay Sorry for the late reply, I have write a tutorial of how to post websites by using IPFS and interact with ENS.

Tutorial link: [https://gist.github.com/PhyrexTsai/cffcbfa1d752b9cf817d920dfcd1ec9f](https://gist.github.com/PhyrexTsai/cffcbfa1d752b9cf817d920dfcd1ec9f)

And also have some website already deployed
- http://phyrextsai.eth
- http://christophershen.eth
- http://hozt.portalnetworkweb.eth
- http://teamnull.bluzellehackathon.eth
- http://cosmos.bluzellehackathon.eth
- http://dturker.bluzellehackathon.eth

PhyrexTsai commented on app/scripts/background.js:
```diff
@@ -154,6 +158,7 @@ async function initialize () {
   const initLangCode = await getFirstPreferredLangCode()
   await setupController(initState, initLangCode)
   log.debug('MetaMask initialization complete.')
```
For the first time can not connect with, the issue that the `initState.NetworkController.provider` is `undefined`, so I think we can have a default value for this issue.

The fix code look likes:
```javascript
let defaultProvider = initState.NetworkController === undefined ? { type: "mainnet" } : initState.NetworkController.provider;
ipfsContent(defaultProvider);
```

danfinlay commented:
The guide looks great @PhyrexTsai. Would you mind if we reposted that on our blog when it's time to publish?

PhyrexTsai commented:
@danfinlay Sure, that will be great, thank you very much

kumavis commented:
we're cutting a bug fix release tomorrow, then this will go in after with the next batch of larger features 😸
thank you for your patience

Arachnid commented:
Any way to make this resolve to the IPFS content, rather than redirecting to it?

pimato commented:
@PhyrexTsai  @danfinlay @kumavis thanks for creating/adding this very important feature. Can´t wait to move the oasis products to IPFS/ENS.

alexvandesande commented:
@Arachnid does a plugin has the right to change DNS hosts and actually register a "eth" TLD, so it looks like a traditional address? Honestly it would be great, but I would understand if google didn't permitted it as it would make phishing powers of extensions much worse

bgits commented:
Shouldn't this feature use IPLD instead of IPFS? The `content` field could be a hash pointing to content stored in all kinds of protocols not just IPFS.

Georgi87 commented:
Are pathnames forwarded? E.g. if I go to mywebsite.eth/home it should redirect to <ipfs-redirect>/home

Event better would be to resolve IPFS content directly. Is this possible to do in a chrome extension @PhyrexTsai ?

decanus commented:
From what I’ve seen, you can resolve content directly if you register a new scheme. So like ipfs:// instead of http://

PhyrexTsai committed:
Update to support pathnames on IPFS hash

As @Georgi87 mentioned about pathnames on the domain like `http://phyrextsai.eth/home` will pass to IPFS hash and append `/home` on redirection url.

You can try the example below: 
http://phyrextsai.eth/index.html
http://phyrextsai.eth/images/phyrex.jpg

PhyrexTsai commented:
@Georgi87 I have committed the feature to let the mechanism in the pathnames of .eth domain working as the same in the ordinary domain.

danfinlay commented:
A few files are not passing our linting rules. We can either:

- Fix these on this branch.
- Add these files to the lint exceptions list.
- Move these files outside of this repository.

The affected files are:

- `app/scripts/lib/ipfsContent.js`
- `app/scripts/lib/contracts/registrar.js`

These errors can be seen here:
https://circleci.com/gh/MetaMask/metamask-extension/31652?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Or by running: `npm run lint`.

These may be automatically fixable by running `npm run lint:fix`.

brunobar79 commented:
I'm gonna submit a PR to this branch fixing the linter error

brunobar79 committed:
Merge remote-tracking branch 'upstream/develop' into HEAD

brunobar79 committed:
fix lint errors

PhyrexTsai committed:
Merge pull request #4 from brunobar79/portal-metamask-lint-fix

Lint fix for Integration ENS with IPFS

PhyrexTsai commented:
@danfinlay @brunobar79 Thanks for helping, I've merged the PR on this branch.

danfinlay reviewed (approved):
Let's do this!

jdetychey commented:
can't wait for this :rocket:

Arachnid commented:
Note that not all ENS names will end in .eth. Can you make this work for other TLDs too?

bgits commented:
@Arachnid What are the planned TLD's or how will new TLD's be proposed and added?

Arachnid commented:
@bgits .luxe and .dds are likely to be  added in the very near future, and correspond to their DNS equivalents. In the longer run, nearly any DNS TLD will be supported via DNSSEC proofs.

mryellow commented on app/manifest.json:
```diff
@@ -59,7 +59,10 @@
+    "https://*.infura.io/",
+    "activeTab",
+    "webRequest",
+    "*://*.eth/"
```
This seems to appear in permissions dialog for Chrome as "eth websites" without an indication it's talking about a TLD. Leaving the user wondering "What is an eth website?"
